clippy_lints/src/matches/single_match.rs

```diff
@@ -140,74 +140,128 @@ fn check_opt_like<'a>(
-    ];
-
-    // We want to suggest to exclude an arm that contains only wildcards or forms the exhaustive
-    // match with the second branch, without enum variants in matches.
-    if !contains_only_wilds(arms[1].pat) && !form_exhaustive_matches(arms[0].pat, arms[1].pat) {
-        return;
+    if check_exhaustive(cx, arms[0].pat, arms[1].pat, ty) {
+        report_single_pattern(cx, ex, arms, expr, els);
     }
+}
 
-    let mut paths_and_types = Vec::new();
-    if !collect_pat_paths(&mut paths_and_types, cx, arms[1].pat, ty) {
-        return;
+/// Resturns true if the given type is one of the standard `Enum`s we know will never get any more
+/// members.
+fn is_known_enum(cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
+    const CANDIDATES: &[&[&str; 3]] = &[&paths::COW, &paths::OPTION, &paths::RESULT];
+    CANDIDATES.iter().any(|&ty_path| match_type(cx, ty, ty_path))
+}
+
+fn contains_only_known_enums(cx: &LateContext<'_>, pat: &Pat<'_>) -> bool {
+    match pat.kind {
+        PatKind::Path(..) => is_known_enum(cx, cx.typeck_results().pat_ty(pat)),
+        PatKind::TupleStruct(..) => {
+            let ty = cx.typeck_results().pat_ty(pat);
+            (contains_only_wilds(pat) || contains_single_binding(pat)) && is_known_enum(cx, ty)
+        },
+        _ => false,
     }
+}
 
-    let in_candidate_enum = |path_info: &(String, &TyS<'_>)| -> bool {
-        let (path, ty) = path_info;
-        for &(ty_path, pat_path) in candidates {
-            if path == pat_path && match_type(cx, ty, ty_path) {
-                return true;
-            }
-        }
-        false
-    };
-    if paths_and_types.iter().all(in_candidate_enum) {
-        report_single_pattern(cx, ex, arms, expr, els);
+fn peel_subpatterns<'a>(pat: &'a Pat<'a>) -> &'a Pat<'a> {
+    if let PatKind::Binding(.., Some(sub)) = pat.kind {
+        return peel_subpatterns(sub);
     }
+    pat
 }
 
-/// Collects paths and their types from the given patterns. Returns true if the given pattern could
-/// be simplified, false otherwise.
-fn collect_pat_paths<'a>(acc: &mut Vec<(String, Ty<'a>)>, cx: &LateContext<'a>, pat: &Pat<'_>, ty: Ty<'a>) -> bool {
-    match pat.kind {
-        PatKind::Wild => true,
-        PatKind::Tuple(inner, _) => inner.iter().all(|p| {
-            let p_ty = cx.typeck_results().pat_ty(p);
-            collect_pat_paths(acc, cx, p, p_ty)
-        }),
-        PatKind::TupleStruct(ref path, ..) => {
-            let path = rustc_hir_pretty::to_string(rustc_hir_pretty::NO_ANN, |s| {
-                s.print_qpath(path, false);
-            });
-            acc.push((path, ty));
-            true
+/// Returns false if the patterns exhaustively match an enum.
+fn check_exhaustive<'a>(cx: &LateContext<'a>, left: &Pat<'_>, right: &Pat<'_>, ty: Ty<'a>) -> bool {
+    let left = peel_subpatterns(left);
+    let right = peel_subpatterns(right);
+    match (&left.kind, &right.kind) {
+        (PatKind::Wild, _) | (_, PatKind::Wild) => true,
```

Comment:
Wouldn't this work?
```suggestion
        (PatKind::Wild | PatKind::Binding(..), _) | (_, PatKind::Wild | PatKind::Binding(..)) => true,
```

Reply:
No, because we should lint bindings iff the `else` branch (`right` pattern) contains wildcards only.